Support for the standard indices() method on slice objects.

nickovs opened the pull request:
Instances of the `slice` class are passed to `__getitem__()` on objects when the user indexes then with a slice. In practice the only thing that anyone ever really wants to do with the `slice` object (other than passing it on untouched) is work out what it means in the context of an array dimension of a particular length. Since Python 2.3 there has been a method on the `slice` class, `indices()`, that takes a dimension length and returns the real `start`, `stop` and `step`, accounting for missing or negative values in the slice spec. This PR implements the `indices()` method on the slice class.

The motivation for implementing this, aside from compatibility, is that the slice syntax provides a convenient and concise way to represent some subset of the regularly spaces input sensors or output devices. I found myself reimplementing the functionality in Python several times for things like sampling infrared thermal sensors or setting patterns on LED arrays. Correctly handling the defaults for `None` values, negative strides, and index values outside beyond the ends requires care to ensure compatibility with CPython.

The implementation here adds 272 bytes (0.02%) to the code for the ESP32. It can be enabled by defining `MICROPY_PY_BUILTINS_SLICE_INDICES` as `1` in `mpconfigport.h`. The PR sets this for the `unix` and `esp32` ports but defaults to `0` for other ports. It should probably be set for other ports that typically have larger ROM capacity.

nickovs committed:
Added support for indices() method on slice objects.

nickovs committed:
Added defautlt value for MICROPY_PY_BUILTINS_SLICE_INDICES in mpconfig.h

nickovs committed:
Fixed cut-and-paste typo

nickovs committed:
Removed debug code

nickovs commented:
Looking into this a bit more, it probably makes sense to expose a C API for doing this, make the `.indices()` method a wrapper around that and then refactor a couple of places where this functionality is duplicated (listed below). This should result in a net reduction in size even with the addition of the functionality.

I'll take a look at doing this and resubmit.

https://github.com/micropython/micropython/blob/973f68780d2eb974e4921ffdf513079efc19e0a4/py/sequence.c#L51

https://github.com/micropython/micropython/blob/973f68780d2eb974e4921ffdf513079efc19e0a4/py/objstrunicode.c#L187

dpgeorge commented on py/objslice.c:
```diff
@@ -53,29 +55,114 @@ STATIC void slice_print(const mp_print_t *print, mp_obj_t o_in, mp_print_kind_t
+#if MICROPY_PY_BUILTINS_SLICE_INDICES
+    } else if (attr == MP_QSTR_indices) {
+        dest[0] = mp_obj_new_bound_meth(MP_OBJ_FROM_PTR(&slice_indices_obj),
+                                        self_in);
```
To avoid allocation of a bound-method object this can be optimised to:
```c
dest[0] = MP_OBJ_FROM_PTR(&slice_indices_obj);
dest[1] = self_in;
```

nickovs replied:
OK, thanks for that. I wasn't quite sure how to do that from the `slice_attr` function.

nickovs committed:
Updated py/objslice.c to expose a new mp_obj_slice_indices() function.

Updated py/obj.h to provide a header for mp_obj_slice_indices() and updated
py/sequence.c and py/objstrunicode.c to use the centralised implementation.

Added new tests for slice indices and for slicing unicode strings.

nickovs committed:
Fixed slice indices tests to correctly be skip when feature is disabled.

nickovs committed:
Removed mp_obj_slice_get() since it is no longer needed (or tested).

nickovs commented:
@dpgeorge With the refactoring in `py/sequence.c` and `py/objstrunicode.c` to use the new, centralised slice indices resolution code the ROM image size for the ESP32 goes _down_ by 48 bytes with the `.indices()` method in place. I have removed the old `mp_obj_slice_get()` C function since it was no longer used.

Given that the new code is actually smaller, should I remove the conditional nature of this code and simply have it in place whenever `MICROPY_PY_BUILTINS_SLICE` is defined?

dpgeorge commented on py/objstrunicode.c:
```diff
@@ -183,28 +183,23 @@ STATIC mp_obj_t str_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
-            mp_obj_slice_get(index, &ostart, &ostop, &ostep);
-            if (ostep != mp_const_none && ostep != MP_OBJ_NEW_SMALL_INT(1)) {
+            mp_bound_slice_t slice;
+            mp_obj_slice_indices(index, utf8_charlen(self_data, self_len), &slice);
```
This call to `utf8_charlen` is going to unconditionally scan the whole (unicode) string on each slice access, to compute the character length.  Note that the original code here, with the calls to `str_index_to_ptr`, was optimised to not scan for the (common) case of small positive start index and small negative end index.  Probably the best/easiest thing to do here is just not touch this part of the code.

nickovs replied:
Yes, I spotted that. This is however a time/space trade-off. If I put it back to the old way then not only is the code larger but I need to reinstate `mp_obj_slice_get()` (which is otherwise unneeded), so we end up with quite a bit more code. Do you have hard statistics for the distribution of the offset sizes used in slices or just a hunch?

dpgeorge replied:
> Do you have hard statistics for the distribution of the offset sizes used in slices or just a hunch?

No.  I was just going by the existing comments in `str_index_to_ptr`.  But the change here would mean that something like `unicode_str[:-1]` will scan the unicode bytes twice, once to determine the number of characters, and once to count len-1 characters.

Yes it's a trade off, but IMO it seems good to keep the current optimised behaviour of this unicode indexing.

Maybe `mp_obj_slice_get()` can be just inlined here instead?

nickovs replied:
OK. I reversed out the refactoring in `str_subscr()` and then in-lined the access to the slice components.

The net result is that with the addition of support for the `.indices()` method, the exposed C API for the `indicies()` functionality, the refactored code in `sequence.c` and the in-lined slice fetch in `str_subscr()`, the whole thing is 32 bytes *smaller* than before I started on the ESP32.

The remaining question is: since this PR both implements a missing Python feature and seems to make the code smaller, should I make it unconditional?

nickovs committed:
Reversed out refactoring of objstrunicode.c for older, more efficient code.

nickovs committed:
In-lined fetching of slice paramters in str_subscr()

dpgeorge commented:
> The remaining question is: since this PR both implements a missing Python feature and seems to make the code smaller, should I make it unconditional?

If it does make the code smaller then I'm happy to have it enabled without a conditional.  But from my measurements I find that this PR (with the feature always enabled when MICROPY_PY_BUILTINS_SLICE is enabled) changes code size by:
```
   bare-arm:    +0 +0.000% 
minimal x86:    +0 +0.000% 
   unix x64:  +312 +0.063% [incl +32(data)]
unix nanbox:  +400 +0.091% [incl +32(data)]
      stm32:  +124 +0.034% PYBV10
     cc3200:  +128 +0.069% 
    esp8266:  +168 +0.026% GENERIC
      esp32:   +76 +0.007% GENERIC[incl +16(data)]
        nrf:  +192 +0.132% pca10040
       samd:  +124 +0.122% ADAFRUIT_ITSYBITSY_M4_EXPRESS
```
I didn't look at it in detail, why the changes are large.

nickovs commented:
Which version are measuring against? Before I started the application part of my ESP32 build was coming out at 1,347,568 bytes and the final version comes in at 1,347,536. If it's making the code grow by more than a trivial amount then it should stay as a conditional, hopefully enabled on platforms with larger ROM sized.

dpgeorge commented:
> Which version are measuring against?

I was using IDF v3 for the builds.  Using IDF v4, this PR does indeed decreases esp32 firmware by 32 bytes.  Strange... it could be because the esp32 port is built with debugging enabled, and this PR looks like it removes an assert (previously in `mp_obj_slice_get()`), which could account for the decrease.

nickovs commented:
For reference, my master is https://github.com/micropython/micropython/commit/973f68780d2eb974e4921ffdf513079efc19e0a4 and the ESP32 build ends saying:
```
Create build-GENERIC/firmware.bin
bootloader     21200
partitions      3072
application  1347568
total        1413104
```

The last commit I have is https://github.com/nickovs/micropython/commit/c94013219fb12a735b87118e8013b99c070e4998 and the build ends with:
```
Create build-GENERIC/firmware.bin
bootloader     21200
partitions      3072
application  1347536
total        1413072
```

dpgeorge commented:
I see that the slice logic you implemented in `mp_obj_slice_indices` is somewhat different to that which existed in (and was removed from) `mp_seq_get_fast_slice_indexes`.  Does your logic change behaviour of this slicing, or is it just expressed in a different way (I guess the latter)?

nickovs commented:
The result of the logic should be the same; I was just trying to make it as succinct as possible. That said, I modelled my logic on (and tested against) CPython, so if your old code had a bug then I might have inadvertently fixed it! :-)

nickovs committed:
py/objslice.c: Replaced mp_obj_int_get_checked() with mp_obj_get_int()

nickovs committed:
Updated mp_bound_slice_t to reflect the fact that it can contain negative values.

nickovs commented:
Is there some action that is needed on my behalf to make this ready for merging?

dpgeorge commented:
> Is there some action that is needed on my behalf to make this ready for merging?

I was hoping a way would be found to reduce code size of the feature so it could be enabled unconditionally (it's possibly more useful than having access to the raw `start`/`stop`/`step` values).  But since there wasn't a way we should keep this conditional.

dpgeorge commented:
Ok, I had some time to look in detail at this.  I like the new `mp_obj_slice_indices()` function, it's clearer to read than the one it replaces and factors out this functionality.  It also looks to be equivalent to the old one in how it works.  I couldn't find any way to further reduce code size.

The commits here were rebased, squashed in to 2 commits, some minor code-formatting changes made, and merged in 4c93955b7b4d3d860aed1551ca6231ac4e388e69 and 10709846f38f8f6519dee27694ce583926a00cb9

Thanks!

nickovs commented:
Glad to help. It's good to have this merged in since now [`uumpy`](https://github.com/nickovs/uumpy) should be able to build with the master branch. Thanks!
